test/test_common/logging.h

```diff
@@ -61,30 +61,103 @@ class LogRecordingSink : public Logger::SinkDelegate {
-        << "\n Actual Log:         " << recorded_log << "\n Expected Substring: " << substr;       \
+    for (const auto& message : log_recorder.messages()) {                                          \
+      if ((message.find(substr) != std::string::npos) &&                                           \
+          (message.find(loglevel) != std::string::npos)) {                                         \
+        FAIL() << "\nLogs:\n"                                                                      \
+               << absl::StrJoin(log_recorder.messages(), "") << "\n Should do NOT contain:\n '"    \
+               << loglevel << "', '" << substr "'\n";                                              \
+      }                                                                                            \
+    }                                                                                              \
+  } while (false)
+
+// Validates that when stmt is executed, exactly one log message containing substr will be emitted.
+// Failure message e.g.,
+//
+// Logs:
+// [2018-04-12 05:51:00.245][7290192][debug][upstream] grpc_mux_impl.cc:160] Received gRPC
+// [2018-04-12 05:51:00.246][7290192][trace][upstream] grpc_mux_impl.cc:80] Sending foo
+//  Does NOT contain:
+//    'warning', 'Too many sendDiscoveryRequest calls for baz’
+#define EXPECT_LOG_CONTAINS(loglevel, substr, stmt)                                                \
+  do {                                                                                             \
+    const ExpectedLogMessages message{{loglevel, substr}};                                         \
+    EXPECT_LOG_CONTAINS_MANY(message, stmt);                                                       \
   } while (false)
 
 // Validates that when stmt is executed, no logs will be emitted.
+// Expected equality of these values:
+//   0
+//   logs.size()
+//     Which is: 3
+//  Logs:
+//   [2018-04-12 05:51:00.245][7290192][debug][upstream] grpc_mux_impl.cc:160] Received gRPC
+//   [2018-04-12 05:51:00.246][7290192][trace][upstream] grpc_mux_impl.cc:80] Sending foo
 #define EXPECT_NO_LOGS(stmt)                                                                       \
   do {                                                                                             \
     LogLevelSetter save_levels(spdlog::level::trace);                                              \
     LogRecordingSink log_recorder(Logger::Registry::getSink());                                    \
     stmt;                                                                                          \
     const std::vector<std::string> logs = log_recorder.messages();                                 \
-    ASSERT_EQ(0, logs.size()) << " Logs:\n   " << absl::StrJoin(logs, "\n  ");                     \
+    ASSERT_EQ(0, logs.size()) << " Logs:\n   " << absl::StrJoin(logs, "   ");                      \
```

Comment:
why did we lose the newline in the error message here?

Reply:
OK I didn't notice they did, but maybe that's because I tested only with a single log.  What you have is good then.